### PR TITLE
Kemal's content_for and yield_content implementation

### PR DESCRIPTION
**Pull request by yorci**
### Description of the Change

Kemal's content_for and yield_content macro added. 
http://kemalcr.com/guide/#content_for-and-yield_content

### Alternate Designs

Laravel's blade implementation will be more flexible but im not skilled about macros :smile: 

### Benefits

Makes changes on layout easier, you dont need multiple layout version for little of part changes  
 

fixes #808

### CODE REVIEW

File: src/amber/controller/helpers/render.cr [hunk 1]
```diff
@@ -15,8 +35,10 @@ module Amber::Controller::Helpers
       {% else %}
         {{ short_path = folder.gsub(/^.+?(?:controllers|views)\//, "") }}
         {% if folder.id.ends_with?(".ecr") %}
+          __content_filename__ = "#{{{path}}}/#{{{short_path.gsub(/\/[^\.\/]+\.ecr/, "")}}}/#{{{filename}}}" # -- 
           %content = render_template("#{{{path}}}/#{{{short_path.gsub(/\/[^\.\/]+\.ecr/, "")}}}/#{{{filename}}}")
         {% else %}
+          __content_filename__ = "#{{{path}}}/#{{{short_path.gsub(/\_controller\.cr|\.cr/, "")}}}/#{{{filename}}}" # -- 
           %content = render_template("#{{{path}}}/#{{{short_path.gsub(/\_controller\.cr|\.cr/, "")}}}/#{{{filename}}}")
         {% end %}
       {% end %}
```
Comment (faustinoaq) [inline src/amber/controller/helpers/render.cr]: I tried this and doesn't work, I think you remove these `__content_filename__` lines and add this:

```cr
macro render(template = nil, layout = true, partial = nil, path = "src/views", folder = __FILE__)
  __content_filename__ = __FILE__

  {% if !(template || partial) %}
  raise "Template or partial required!"
  {% end %}

  {{ filename = template || partial }}


  # Render Template and return content
  {% if filename.id.split("/").size > 1 %}
    %content = render_template("#{{{filename}}}", {{path}})
  {% else %}
    {{ short_path = folder.gsub(/^.+?(?:controllers|views)\//, "") }}
    {% if folder.id.ends_with?(".ecr") %}
      %content = render_template("#{{{path}}}/#{{{short_path.gsub(/\/[^\.\/]+\.ecr/, "")}}}/#{{{filename}}}")
    {% else %}
      %content = render_template("#{{{path}}}/#{{{short_path.gsub(/\_controller\.cr|\.cr/, "")}}}/#{{{filename}}}")
    {% end %}
  {% end %}

  # Render Layout
  {% if layout && !partial %}
  content = %content
  render_template("#{{{path}}}/layouts/#{{{layout.class_name == "StringLiteral" ? layout : LAYOUT}}}")
  {% else %}
  %content
  {% end %}
end
```

^^ This way works just fine :tada: Maybe @sdogruyol has some ideas here :wink:

**Comment by faustinoaq** [comment]
BTW, If we gonna add this would be nice a new section here: https://amberframework.gitbook.io/amber/guides/views

screenshot_20180521_073702


We have an **Edit on Github** button now, so feel free to contribute opening more issues and even PRs :wink:

File: src/amber/controller/helpers/render.cr (hunk 1 above)
Comment (yorci) [reply src/amber/controller/helpers/render.cr]: Well thats not works for me, i just edit yiled_content and add this line 
```
puts __content_filename__,  CONTENT_FOR_BLOCKS[{{key}}][0]
```

And,

when render homepage template the output is 
src/views/home/homepage.ecr - __content_filename__
src/views/home/homepage.ecr - CONTENT_FOR_BLOCKS[{{key}}][0]

when render login template the output is 
/home/ysr/***/***/***/***/src/controllers/auth_controller.cr - __content_filename__
src/views/auth/register.ecr  - CONTENT_FOR_BLOCKS[{{key}}][0]


can you try it again?

**Comment by sdogruyol** [comment]
Nice to see my implementation being helpful to others :+1:

File: src/amber/controller/helpers/render.cr (hunk 1 above)
Comment (faustinoaq) [reply src/amber/controller/helpers/render.cr]: Still testing, sorry for the delay :sweat_smile:

Comment (faustinoaq) [reply src/amber/controller/helpers/render.cr]: BTW, this PR need to solve some merge conflicts and formatting issues :+1:

**Comment by faustinoaq** [comment]
Hi @elorest There are some merging issues here, related to #791 Can you review this? :sweat_smile:

**Comment by faustinoaq** [comment]
Hi @yorci I just tried to fix merge conflicts, Can you test my changes? :sweat_smile:

**Comment by faustinoaq** [comment]
@yorci No problem, Thank you so much for open this PR! :100: 

We'll add specs to it then :wink: :+1:

**Comment by yorci** [comment]
@elorest  sorry for late return, thats the Kemal's docs link http://kemalcr.com/guide/#content_for-and-yield_content

**Pull request closed without merging**